src/lib/storage/table.cpp

```diff
@@ -373,39 +403,114 @@ void Table::create_chunk_index(const std::vector<ColumnID>& column_ids, const st
+
+  // Check validity of key columns from other table.
+  const auto referenced_table_column_count = referenced_table->column_count();
+  for (const auto& column_id : foreign_key_constraint.primary_key_columns()) {
+    Assert(column_id < referenced_table_column_count, "ColumnID out of range.");
+  }
+
+  const auto append_lock = acquire_append_mutex();
+  for (const auto& existing_constraint : _foreign_key_constraints) {
+    // Do not allow overlapping foreign key constraints. Though a table may have unlimited inclusion dependencies for
+    // the same columns (and especially the existence of [a] in [x] and [b] in [y] does not mean [a, b] in [x, y]
+    // holds), it is reasonable to assume disjoint (soft) foreign keys for now.
+    Assert(!columns_intersect(existing_constraint.foreign_key_columns(), foreign_key_constraint.foreign_key_columns()),
+           "ForeignKeyConstraint for required columns has already been set.");
+  }
+  _foreign_key_constraints.insert(foreign_key_constraint);
+  const auto referenced_table_append_lock = referenced_table->acquire_append_mutex();
+  referenced_table->_referenced_foreign_key_constraints.insert(foreign_key_constraint);
+}
+
+const ForeignKeyConstraints& Table::soft_foreign_key_constraints() const {
+  return _foreign_key_constraints;
+}
+
+const ForeignKeyConstraints& Table::referenced_foreign_key_constraints() const {
+  return _referenced_foreign_key_constraints;
+}
+
+void Table::add_soft_order_constraint(const TableOrderConstraint& table_order_constraint) {
+  Assert(_type == TableType::Data, "TableOrderConstraints are not tracked for reference tables across the PQP.");
+
+  // Check validity of columns.
+  const auto column_count = this->column_count();
+  for (const auto& column_id : table_order_constraint.ordering_columns()) {
+    Assert(column_id < column_count, "ColumnID out of range.");
   }
+
+  // Check validity of ordered columns.
+  for (const auto& column_id : table_order_constraint.ordered_columns()) {
+    Assert(column_id < column_count, "ColumnID out of range.");
+  }
+
+  const auto append_lock = acquire_append_mutex();
+  for (const auto& existing_constraint : _table_order_constraints) {
+    // Do not allow order constraints that can be represented by existing constraints (and vice versa). For instance,
+    // we do not allow constraints that (i) order the same ordered columns but require fewer or more ordering columns
+    // or (ii) have the same ordering columns but order fewer or more ordered columns.
+    const auto ordering_columns_invalid =
+        columns_overlap(existing_constraint.ordering_columns(), table_order_constraint.ordering_columns()) &&
+        existing_constraint.ordered_columns() == table_order_constraint.ordered_columns();
+    const auto ordered_columns_invalid =
+        columns_overlap(existing_constraint.ordered_columns(), table_order_constraint.ordered_columns()) &&
```

Comment:
See question above. Adding `[1,2]` would not fail in `[0,1,2]` is already in place, right?

Reply:
You're right and I really had to think this through. What basically happens if both your OD examples are valid is that there is a transitive OD: if `[a,b] |-> [x]` and `[b] |-> [x]`, also `[a] |-> [b]` must be valid, and hence, also `[a] |-> [c]`. So we should not look from the beginning here but ensure there are no intersections (might be to pessimistic, but should work for now). I think this would at least make some noise if we ever run into an intricate case

Reply:
Simplified to fail if columns intersect.